File: 지찬/프로그래머스 코테 kit/신고결과받기.java
//https://school.programmers.co.kr/learn/courses/30/lessons/92334
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

class Solution {

    HashMap<String, ArrayList<String>> reporters = new HashMap<>();
    HashMap<String, Integer> reported_times = new HashMap<>();

    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];
        for (String id :
                id_list) {
            reporters.put(id, new ArrayList<>());
            reported_times.put(id, 0);
        }

        for (String r :
                report) {
            String[] s = r.split(" ");
            String reporter = s[0];
            String reported = s[1];
            ArrayList<String> arrayList = reporters.get(reported);
            if(!arrayList.contains(reporter)) {
                arrayList.add(reporter);
                reported_times.put(reported, reported_times.get(reported) + 1);
            }
        }

        for (String key : reported_times.keySet()) {
            if (reported_times.get(key) >= k) {
                for (String id :
                        reporters.get(key)) {
                    for (int i = 0; i < id_list.length; i++) {
                        if (id_list[i].equals(id)) {
                            answer[i]++;
                            break;
                        }
                    }
                }
            }
        }
        return answer;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String[] id_list = {
                "muzi", "frodo", "apeach", "neo"
        };
        String[] report = {
                "muzi frodo", "apeach frodo", "frodo neo", "muzi neo", "apeach muzi"
        };
        int k = 2;
        System.out.println(Arrays.toString(solution.solution(id_list, report, k)));
    }
}

File: 성민/레벨1/신고_결과_받기.js
function solution(id_list, report, k) {
  const result = [];
  const reporterMap = new Map([]);
  const reportedMap = new Map([]);
  for (const id of id_list) {
    reporterMap.set(id, []);
    reportedMap.set(id, []);
  }

  for (const r of report) {
    const [reporter, reported] = r.split(" ");
    const reporterArray = reporterMap.get(reporter);
    const reportedArray = reportedMap.get(reported);
    if (reporterArray.includes(reported)) continue;
    reporterArray.push(reported);
    reportedArray.push(reporter);
    reporterMap.set(reporter, reporterArray);
    reportedMap.set(reported, reportedArray);
  }

  const blockUsers = [...reportedMap.entries()]
    .filter((v) => !(v[1].length < k))
    .map((v) => v[0]);

  for (const v of reporterMap.values()) {
    let sendMailCnt = 0;
    for (const r of blockUsers) {
      if (v.includes(r)) sendMailCnt++;
    }
    result.push(sendMailCnt);
  }
  return result;
}

solution(
  ["muzi", "frodo", "apeach", "neo"],
  ["muzi frodo", "apeach frodo", "frodo neo", "muzi neo", "apeach muzi"],
  2
);
